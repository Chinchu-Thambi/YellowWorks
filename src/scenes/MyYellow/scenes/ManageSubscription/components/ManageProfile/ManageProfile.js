import React from 'react';
import PropTypes from 'prop-types';

import Button from '../../../../../../components/Button';

import profileIcon from '../../../../../../assets/icons/profile.svg';

import {
  getFirstPricingPlan, formatPricingPlanAmount, // getTotalFromSubscription,
} from '../../../../../../services/getPricingPlans';

import {
  StyledSection, FormControlSelect, DowngradeWarning,
} from '../../Styled';
import ProfileTypes from '../modals/Profiles';

const shortenName = (name, section) => name.split(' ')?.[section || 0];

const ManageProfile = (props) => {
  const {
    renderModalContent,
    handleProfileChange,
    selectableProducts,
    status,
    subscriptionSummary,
    mergedSubscription,
    disableInputs,
  } = props;

  return (
    <StyledSection justifyContent="space-between">
      <h2>
        <img src={profileIcon} alt="" />
        PROFILE
      </h2>
      <Button
        variant="link"
        size="sm"
        onClick={() => renderModalContent({
          title: 'Profile Types',
          component: () => <ProfileTypes />,
        })}
      >
        Learn More
      </Button>
      <FormControlSelect
        id="profile"
        aria-label="Profile Type"
        disabled={status === 'PENDING' || disableInputs}
        value={subscriptionSummary?.upcomingOrderDetails?.product?.sku || mergedSubscription?.product?.sku}
        onChange={handleProfileChange}
        required
      >
        <option value="" disabled>Choose Profile</option>
        {selectableProducts?.map((profile) => {
          const pricingPlan = profile?.pricingPlan?.[1] || getFirstPricingPlan(profile);
          return (
            <option
              key={profile.sku}
              value={profile.sku}
            >{formatPricingPlanAmount(pricingPlan)} - {shortenName(profile.name, 1)}
            </option>
          );
        })}
      </FormControlSelect>
      {(subscriptionSummary?.orderDetails?.product.sku === 'YPPREMIUMLOYAL'
        && mergedSubscription?.product?.sku !== 'YPPREMIUMLOYAL')
        && <DowngradeWarning />}
    </StyledSection>
  );
};

ManageProfile.defaultProps = {
  renderModalContent: () => { },
  handleProfileChange: () => { },
  selectableProducts: [],
  disableInputs: true,
  subscriptionSummary: {},
  mergedSubscription: {},
};

ManageProfile.propTypes = {
  renderModalContent: PropTypes.func,
  handleProfileChange: PropTypes.func,
  selectableProducts: PropTypes.arrayOf(PropTypes.shape({})),
  disableInputs: PropTypes.bool,
  status: PropTypes.string.isRequired,
  subscriptionSummary: PropTypes.shape({
    upcomingOrderDetails: PropTypes.shape({
      product: PropTypes.shape({
        sku: PropTypes.string,
      }),
    }),
    orderDetails: PropTypes.shape({
      product: PropTypes.shape({
        sku: PropTypes.string,
      }),
    }),
  }),
  mergedSubscription: PropTypes.shape({
    product: PropTypes.shape({
      sku: PropTypes.string,
    }),
  }),
};

export default ManageProfile;
