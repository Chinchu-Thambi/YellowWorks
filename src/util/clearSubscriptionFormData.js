/* globals localStorage */

import { storageAvailable } from '.';

const clearSubscriptionFormData = () => {
  if (storageAvailable()) {
    Object.keys(localStorage).forEach((key) => {
      if (
        key === 'storageOrderDetails'
        || key.startsWith('storageFormData')
        || key.startsWith('formData-') // still needed because onboarding uses it, yet.
      ) {
        localStorage.removeItem(key);
      }
    });
  }
};

export default clearSubscriptionFormData;
