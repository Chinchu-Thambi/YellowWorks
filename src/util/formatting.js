const zeroBasedMonths = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

export const monthIndexToString = (monthIndex) => zeroBasedMonths[monthIndex];

/**
 * Returns date of format `MMM 'yy` e.g. `Jun '19`
 * @param date
 */
export const formatDateToMonthYear = (date) => `${monthIndexToString(date.getMonth())} '${date.getFullYear().toString().substr(-2)}`;

/**
 * Returns date of format 'dd MMM yyyy' e.g. '2 Jun 2019'
 * @param date
 */
export const formatDateWithMonthName = (date) => `${date.getDate()} ${monthIndexToString(date.getMonth())} ${date.getFullYear().toString()}`;

const padNumber = (number) => (number < 10 ? `0${number}` : `${number}`);

/**
 * Returns `YYYYMM` e.g. August 2020 would be `2020-08`
 */
export const formatYYYYdashMM = (date) => `${date.getFullYear()}-${padNumber(date.getMonth() + 1)}`;

/**
 * Returns `YYYYMM` e.g. August 2020 would be `202008`
 */
export const formatYYYYMM = (date) => `${date.getFullYear()}${padNumber(date.getMonth() + 1)}`;

export const toSentenceCase = (str) => {
  if (!str) return str;

  const newStr = str.toLowerCase();
  const firstLetterCapitalized = newStr.substring(0, 1).toUpperCase();

  if (newStr.length > 1) {
    return firstLetterCapitalized + newStr.substring(1);
  }
  return firstLetterCapitalized;
};

export const utcDate = (year, month0based, day1based) => new Date(Date.UTC(year, month0based, day1based));
