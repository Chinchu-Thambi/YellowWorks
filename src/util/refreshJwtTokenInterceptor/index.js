import refreshJwtToken from './refreshJwtTokenInterceptor';

export default refreshJwtToken;
