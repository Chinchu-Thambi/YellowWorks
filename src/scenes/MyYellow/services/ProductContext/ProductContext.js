import React from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import * as R from 'ramda';
import { NotificationManager } from 'react-notifications';

import AuthContext from '../../../../components/Auth/AuthContext';
import { useLocalStorage } from '../../../../util';

const ProductContext = React.createContext();
export default ProductContext;

export const ProductProvider = ({ children }) => {
  const { jwtToken, customerId } = React.useContext(AuthContext) || {};
  const [subscriptionId, setSubscriptionId] = React.useState(null);
  const [storageFormData, setStorageFormData] = useLocalStorage('storageFormData', null);
  const [storageOrderDetails, setStorageOrderDetails] = useLocalStorage('storageOrderDetails', null);
  const [storageFormDataExpiry, setStorageFormDataExpiry] = useLocalStorage('storageFormDataExpiry', {});
  const [isLoading, setIsLoading] = React.useState(false);
  const [storageErrors, setStorageErrors] = useLocalStorage('storageFormDataErrors', null);
  const formData = storageFormData?.[subscriptionId] || null;
  const orderDetails = storageOrderDetails?.[subscriptionId] || null;
  const expiryTime = storageFormDataExpiry?.[subscriptionId] || null;
  const errors = storageErrors?.[subscriptionId] || null;

  const expireFormData = React.useCallback(() => {
    delete storageFormDataExpiry[subscriptionId];
    setStorageFormDataExpiry({
      ...storageFormDataExpiry,
    });
  }, [setStorageFormDataExpiry, storageFormDataExpiry, subscriptionId]);

  const setExpiryDate = React.useCallback((delay) => {
    const now = new Date();
    const futureDate = new Date(now.getTime() + delay).toISOString();
    setStorageFormDataExpiry({
      ...storageFormDataExpiry,
      [subscriptionId]: futureDate,
    });
  }, [setStorageFormDataExpiry, storageFormDataExpiry, subscriptionId]);

  const saveData = async ({ path, data }) => {
    const updatedFormData = R.assocPath(
      path,
      data,
    )(formData);

    if (!R.equals(updatedFormData, formData)) {
      const query = `
        mutation submitSchema(
          $subscriptionId: ID!
          $customerId: ID!
          $input: AWSJSON!
          $publish: Boolean!
        ) {
          updateConfiguration(
            subscriptionId: $subscriptionId
            customerId: $customerId
            input: $input
            publish: $publish
          )
        }
      `;

      const variables = {
        subscriptionId,
        customerId,
        input: JSON.stringify(updatedFormData),
        publish: true,
      };

      const response = await axios.post(
        process.env.GATSBY_PRODUCT_PLATFORM_ENDPOINT,
        { query, variables },
        { headers: { Authorization: jwtToken } },
      );

      const submitProductBrief = R.compose(
        R.tryCatch(JSON.parse, () => ({})),
        R.path(['data', 'data', 'updateConfiguration']),
      )(response);

      if (submitProductBrief.error) {
        console.error('There was a problem with the user input.', {
          key: `subscription_${subscriptionId}`,
          variables,
          orderDetails,
        });
        NotificationManager.error('Input could not be saved.', '', 5000);
        return { success: false };
      }

      setStorageFormData({
        ...storageFormData,
        [subscriptionId]: updatedFormData,
      });

      const in3Seconds = 3 * 1000;
      setExpiryDate(in3Seconds);

      NotificationManager.success('Saved successfully.', '', 5000);

      return { success: true };
    }

    return { success: false };
  };

  // load formData
  React.useEffect(() => {
    // bare requirements for formData loading
    if (
      !jwtToken
      || !subscriptionId
      || isLoading
    ) return;

    // controls for data based loading necessity
    if (
      expiryTime && formData
    ) return;

    const tomorrow = (24 * 60 * 60 * 1000);
    setExpiryDate(tomorrow);
    setStorageErrors({
      ...storageErrors,
      [subscriptionId]: null,
    });

    const getSubscriptionContent = async () => {
      const query = `
        query getProductContextData($id: ID!, $customerId: ID!) {
          subscription(
            id: $id
            customerId: $customerId
          ) {
              id
              orderDetails {
                updated
                activated
                product {
                  sku
                  id
                  name
                  configuration
                  pricingPlan
                }
                productOptions {
                  name
                  id
                  sku
                  configuration
                  pricingPlan
                }
                providerData {
                  slug
                }
                status
                userSchema
              }
              productDetails {
                userConfiguration
                status
              }
              upcomingOrderDetails {
                updated
                activated
                product {
                  sku
                  id
                  name
                  configuration
                  pricingPlan
                }
                productOptions {
                  name
                  id
                  sku
                  configuration
                  pricingPlan
                }
                status
              }
            }
          }
      `;
      const variables = {
        id: subscriptionId,
        customerId,
      };

      if (expiryTime) {
        setStorageFormData({
          ...storageFormData,
          [subscriptionId]: undefined,
        });
        setStorageOrderDetails({
          ...storageOrderDetails,
          [subscriptionId]: undefined,
        });
      }
      setIsLoading(true);

      const response = await axios.post(
        process.env.GATSBY_PRODUCT_PLATFORM_ENDPOINT,
        { query, variables },
        { headers: { Authorization: jwtToken } },
      );

      if (response.data?.errors) {
        setStorageErrors({ [subscriptionId]: response.data.errors });
      }

      const orderDetailsData = R.path(['data', 'data', 'subscription', 'orderDetails'])(response);
      const userSchema = R.tryCatch(JSON.parse, () => {
        setStorageErrors({
          [subscriptionId]: [
            ...response.data.errors,
            {
              errorType: 'Schema parsing',
              message: 'Schema could not be parsed.',
            },
          ],
        });
        return null;
      })(orderDetailsData?.userSchema);
      orderDetailsData.userSchema = userSchema;

      const data = R.compose(
        R.tryCatch(JSON.parse, () => {
          setStorageErrors({
            [subscriptionId]: [
              ...response.data.errors,
              {
                errorType: 'Data parsing',
                message: 'Data could not be parsed.',
              },
            ],
          });
          return null;
        }),
        R.path(['data', 'data', 'subscription', 'productDetails', 'userConfiguration']),
      )(response);

      setStorageFormData({
        ...storageFormData,
        [subscriptionId]: data,
      });
      setStorageOrderDetails({
        ...storageOrderDetails,
        [subscriptionId]: orderDetailsData,
      });
      setIsLoading(false);
    };
    getSubscriptionContent();
  }, [customerId, expiryTime, formData, isLoading, jwtToken, orderDetails, setStorageErrors, setExpiryDate,
    setStorageFormData, setStorageOrderDetails, storageFormData, storageOrderDetails, subscriptionId, storageErrors]);

  // setup reload for existing expiration time
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    const now = new Date();
    const expiryDate = new Date(expiryTime);
    const delay = expiryDate - now;

    const timeout = setTimeout(() => {
      expireFormData();
    }, delay);

    return () => { clearTimeout(timeout); };
  }, [expireFormData, expiryTime]);

  // sets expiration time for pending listing
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    if (orderDetails?.status === 'PENDING') {
      const delay = 5000;
      const timeout = setTimeout(() => {
        expireFormData();
      }, delay);

      return () => { clearTimeout(timeout); };
    }

    return () => {};
  }, [expireFormData, expiryTime, orderDetails]);

  // sets expiration time for when an error happens
  React.useEffect(() => {
    if (!expiryTime) { return () => {}; }

    if (errors) {
      const in20Seconds = 20 * 1000;
      const timeout = setTimeout(() => {
        expireFormData();
      }, in20Seconds);

      return () => { clearTimeout(timeout); };
    }

    return () => {};
  }, [errors, expireFormData, expiryTime]);


  const productState = {
    // status
    subscriptionId,
    errors,

    // data
    formData,
    orderDetails,

    // action
    saveData,
    setSubscriptionId,
    expireFormData,
  };

  return (
    <ProductContext.Provider
      value={productState}
    >{children}
    </ProductContext.Provider>
  );
};

ProductProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
