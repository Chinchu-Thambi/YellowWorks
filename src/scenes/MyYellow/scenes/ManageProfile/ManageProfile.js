import React from 'react';
import PropTypes from 'prop-types';
import { Box } from 'rebass/styled-components';
import styled from 'styled-components';
import { mediaBreakpointUp } from '../../../../util';
import Spinner from '../../../../components/Spinner';

import ProductContext from '../../services/ProductContext';

import Essentials from './components/Essentials'; // @TODO: add this when functionality is ready
import PrimaryInformation from './components/PrimaryInformation';
import ViewYourProfile from './components/ViewYourProfile';
import EditGallery from './components/EditGallery'; // @TODO: add this when functionality is ready
import ManageCategories from './components/ManageCategories';
import MainInfoTab from './components/MainInfoTab';
import ErrorMessage from './components/ErrorMessage';
import AdditionalInfoTab from './components/AdditionalInfoTab';
import CloseBusiness from './components/CloseBusiness';

const FlexContainer = styled.div`
  display: block;
  color: ${({ theme }) => theme.palette.contrast[1]};

  ${mediaBreakpointUp('lg')} {
    display: flex;
  }
`;

const CardContainer = styled(Box)`
  width: 100%;

  /* Second column */
  ${mediaBreakpointUp('lg')} {
      &:last-child {
      max-width: 447px;
    }
  }
`;

const ManageProfile = (props) => {
  const { subscriptionId } = props;

  const {
    errors, formData, setSubscriptionId, orderDetails,
  } = React.useContext(ProductContext) || {};

  // set subscription Id
  React.useEffect(() => {
    if (typeof setSubscriptionId === 'function') {
      setSubscriptionId(subscriptionId);
    }
  }, [setSubscriptionId, subscriptionId]);

  if (errors) return <ErrorMessage errors={errors} />;
  if (!formData) return <Spinner />;

  return (
    <FlexContainer>
      <CardContainer>
        <Essentials />
        <PrimaryInformation />
        <AdditionalInfoTab schema={orderDetails?.userSchema} />
      </CardContainer>

      <CardContainer>
        <ViewYourProfile />
        <EditGallery />
        <ManageCategories />
        <MainInfoTab />
        <CloseBusiness />
      </CardContainer>
    </FlexContainer>
  );
};

ManageProfile.propTypes = {
  subscriptionId: PropTypes.string.isRequired,
};

export default ManageProfile;
