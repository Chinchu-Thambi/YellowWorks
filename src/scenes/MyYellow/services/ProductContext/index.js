import ProductContext, { ProductProvider } from './ProductContext';

export { ProductProvider };
export default ProductContext;
