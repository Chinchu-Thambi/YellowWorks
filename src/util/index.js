/* global window, DOMException, document */

import React from 'react';
import showdown from 'showdown';

import { importableModules } from '../services/useModule';

export {
  default as theme,
  mediaBreakpointUp,
} from './theme';
export {
  default as errorText,
  infos as infoText,
  getErrorString,
} from './texts';

const isSingleLine = (string) => (string !== undefined ? !string.includes('\n') : '');

const mdConverter = new showdown.Converter();

mdConverter.setOption('tables', true);

export const parseMd = (md) => {
  if (!md) {
    return '';
  }

  let mdString = md;

  if (typeof md === 'object') {
    [mdString] = Object.values(md);
  }

  let parsedString = mdConverter.makeHtml(mdString);

  if (isSingleLine(mdString)) {
    // remove outer p tag from single line md
    parsedString = parsedString.replace(/<\/?p>/g, '');
  }

  // eslint-disable-next-line react/no-danger
  return <span dangerouslySetInnerHTML={{ __html: parsedString }} />;
};

export const getProportionalHeight = ({
  initialWidth, initialHeight, targetWidth,
}) => (initialHeight * targetWidth) / initialWidth;

export const storageAvailable = () => {
  let storage;

  if (typeof window !== 'undefined') {
    storage = window.localStorage;
  }

  try {
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    if (typeof DOMException === 'undefined') {
      return false;
    }

    return e instanceof DOMException && (
      // everything except Firefox
      e.code === 22
      // Firefox
      || e.code === 1014
      // test name field too, because code might not be present
      // everything except Firefox
      || e.name === 'QuotaExceededError'
      // Firefox
      || e.name === 'NS_ERROR_DOM_QUOTA_REACHED')
      // acknowledge QuotaExceededError only if there's something already stored
      && storage && storage.length !== 0;
  }
};

export const formatPrice = (price = 0) => (typeof price === 'number' ? `$${price.toFixed(2)}` : `$${price}`);

export const formatRecurringPrice = (price, recurring, withGst, annual) => {
  const priceFormatted = formatPrice(price);
  if (priceFormatted === '$0.00') {
    return 'Free';
  }

  if (recurring) {
    if (withGst) {
      return `${priceFormatted} / month + GST`;
    }
    if (annual) {
      return `${priceFormatted} / year`;
    }
    return `${priceFormatted} / month`;
  }
  if (withGst) {
    return `${priceFormatted} + GST`;
  }
  return priceFormatted;
};

// ex input date object
// ex out: "13/08/2019"
export const reformatDate = (dateObj) => (
  dateObj ? dateObj.toLocaleDateString('en-NZ') : ''
);

export const useLocalStorage = (key, initialValue) => {
  // from https://usehooks.com/useLocalStorage/
  // State to store our value
  // Pass initial state function to useState so logic is only executed once
  const [storedValue, setStoredValue] = React.useState(() => {
    try {
      // Get from local storage by key
      if (storageAvailable()) {
        const item = window.localStorage.getItem(key);
        // Parse stored json or if none return initialValue
        return item ? JSON.parse(item) : initialValue;
      }
      return initialValue;
    } catch (error) {
      // If error also return initialValue
      console.log(error);
      return initialValue;
    }
  });

  // Return a wrapped version of useState's setter function that ...
  // ... persists the new value to localStorage.
  const setValue = (value) => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      if (storageAvailable()) {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
    } catch (error) {
      // A more advanced implementation would handle the error case
      console.log(error);
    }
  };

  return [storedValue, setValue];
};

// TODO: remove this and user version inside of auth/services
export const isStaff = async () => {
  try {
    const { Auth } = await importableModules['@aws-amplify/auth']();
    const user = await Auth.currentSession();
    return user.idToken.payload.email.endsWith('@yellow.co.nz');
  } catch (err) {
    // console.log('isStaff', err);
    return false;
  }
};


export const isTesting = () => document.location.search.includes('utm_isTesting=yes');

export const getNextWeekday = () => {
  const dt = new Date();
  dt.setDate(dt.getDate() + 1);

  if (dt.getDay() === 6) { // if Saturday
    dt.setDate(dt.getDate() + 2);
  }

  if (dt.getDay() === 0) { // if Sunday
    dt.setDate(dt.getDate() + 1);
  }

  return dt;
};

export const getNextMonth = ({
  baseNextWeekday = false,
} = {}) => {
  let dt = new Date();

  if (baseNextWeekday) {
    dt = getNextWeekday();
  }

  dt.setFullYear(
    dt.getFullYear(),
    dt.getMonth() + 1,
    1,
  );

  return dt;
};

export const getEndOfMonth = ({
  baseNextWeekday = false,
} = {}) => {
  let dt = new Date();

  if (baseNextWeekday) {
    dt = getNextWeekday();
  }

  dt.setFullYear(
    dt.getFullYear(),
    dt.getMonth() + 1,
    0,
  );

  return dt;
};

export const capitalize = (string = '') => `${string.charAt(0).toUpperCase()}${string.substr(1).toLowerCase()}`;

export const shuffle = (array) => {
  let counter = array.length;
  const shuffled = [...array];

  // While there are elements in the array
  while (counter > 0) {
    // Pick a random index
    const index = Math.floor(Math.random() * counter);

    // Decrease counter by 1
    counter -= 1;

    // And swap the last element with it
    const temp = shuffled[counter];
    shuffled[counter] = shuffled[index];
    shuffled[index] = temp;
  }

  return shuffled;
};
