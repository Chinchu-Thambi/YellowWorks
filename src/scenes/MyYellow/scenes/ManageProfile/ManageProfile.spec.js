/* eslint-disable no-console */
/* globals describe, it, beforeEach, afterAll, expect, jest */

import React from 'react';
import {
  render, cleanup, fireEvent, act, getByRole,
} from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { theme } from '../../../../util';

import ProductContext from '../../services/ProductContext';

import ManageProfile from './ManageProfile';

describe('ManageProfile', () => {
  const originalConsole = { ...console };

  beforeEach(() => {
    cleanup();
  });

  afterAll(() => {
    cleanup();
  });

  it('does not throw with incomplete data', () => {
    console.error = jest.fn();

    expect(() => {
      render(<ManageProfile />);
    }).not.toThrow();

    // this avoids proptype error log pollution but captures the content
    expect(console.error.mock.calls).toMatchSnapshot();

    // restore console.error
    console.error = originalConsole.error;
  });

  it('implements phone number editing functionality', async () => {
    // set initial state
    const telephone = { number: '32165132', areaCode: '7', countryCode: '64' };

    const productState = {
      formData: {
        location: {
          telephone,
        },
      },
      saveData: jest.fn().mockResolvedValue({ success: true }),
    };

    // render
    const wrapper = render(
      <ThemeProvider theme={theme}>
        <ProductContext.Provider value={productState}>
          <ManageProfile subscriptionId="subscriptionId" />
        </ProductContext.Provider>
      </ThemeProvider>,
    );

    // select by phone number title
    const phoneNumberContainer = wrapper.getByText('Phone Numbers').parentElement.parentElement;
    const phoneNodeContent = phoneNumberContainer.textContent;

    // expect phone number content to be rendered on the screen
    expect(phoneNodeContent).toContain(telephone.areaCode);
    expect(phoneNodeContent).toContain(telephone.number);


    // edit functionality
    const editButton = getByRole(phoneNumberContainer, 'button', { name: 'Edit' });
    editButton.click();

    // define new phone values
    const newPhone = '1111111';
    const newAreaCode = '9';

    const phoneInput = wrapper.getByDisplayValue(telephone.number);
    const areaInput = wrapper.getByDisplayValue(`0${telephone.areaCode}`);

    // async because the second event depends on applied changes from the first.
    await fireEvent.change(phoneInput, { target: { value: newPhone } });
    await fireEvent.change(areaInput, { target: { value: newAreaCode } });

    // expect updated values on the input elements
    expect(phoneInput.value).toBe(newPhone);
    expect(areaInput.value).toBe(newAreaCode);

    // save functionality
    const saveButton = wrapper.getByRole('button', { name: 'Save' });

    await act(async () => {
      saveButton.click();
    });

    expect(productState.saveData.mock.calls[0][0]).toMatchSnapshot();
  });
});
