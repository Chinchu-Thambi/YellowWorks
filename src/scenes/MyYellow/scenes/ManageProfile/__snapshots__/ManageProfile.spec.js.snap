// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageProfile does not throw with incomplete data 1`] = `
Array [
  Array [
    "Warning: Failed prop type: The prop \`subscriptionId\` is marked as required in \`ManageProfile\`, but its value is \`undefined\`.
    in ManageProfile",
  ],
]
`;

exports[`ManageProfile implements phone number editing functionality 1`] = `
Object {
  "data": Object {
    "contactPoint": Array [],
    "mobileNumber": Object {},
    "telephone": Object {
      "areaCode": "9",
      "countryCode": "64",
      "number": "1111111",
    },
    "tollFreeNumber": Object {},
  },
  "path": Array [
    Array [
      "location",
    ],
  ],
}
`;
